README.txt
GM_PHD_Filter
Version 1.09, 13th December 2013

Matlab code by Bryan Clarke b.clarke@acfr.usyd.edu.au with:
- some Kalman filter update code by Tim Bailey, taken from his website http://www-personal.acfr.usyd.edu.au/tbailey/software/
- error_ellipse by AJ Johnson, taken from Matlab Central http://www.mathworks.com.au/matlabcentral/fileexchange/4705-errorellipse
Algorithm by Ba-Ngu Vo & Wing-Kin Ma in:
B.-N. Vo, W.-K. Ma, "The Gaussian Mixture Probability Hypothesis Density Filter", IEEE Transactions on Signal Processing, Vol 54, No. 11, November 2006, pp4091-4104

This is an implementation of a gaussian mixture probability hypothesis density filter (GM-PHD) for a simulated tracking problem. The problem specification is given in the above paper - in summary, two targets move through the environment, there is a lot of clutter on the measurement, and about halfway through a third target spawns off one of the two targets.

I made a few changes, either because I couldn't understand how Vo&Ma did it, or because I wanted to make it closer to my target problem. I extended the measurement vector to include both target position AND velocity (the filter they describe tracks position and velocity but only observes position). Velocity is observed as just being dx/dt, change in position over time, between this new observation and the previous position of this target. Targets are either birthed or spawned depending on which initialisation weight function would give a higher weight; they are given the appropriate initialisation covariance, but apart from this there is no other difference in instantiation. Vo & Ma's state estimation involves repeatedly outputting targets that have a rounded weight greater than 1; I have this off by default but it can be turned on by setting OUTPUT_MULTIPLE_HIGH_WEIGHT_TARGETS and VERBOSE to 1 in GM_PHD_Initialisation.

I've also implemented the Extended Kalman Filter version specified by Vo & Ma, but not the EKF simulation. To switch from linear KF to EKF, set USE_KF in GM_PHD_Initialisation to 1. Everything runs out of the same file, GM_PHD_Filter. The EKF uses either totally separate or totally identical files to the linear version. I considered intermingling the EKF code into the KF using a bunch of IF statements to make maintenance easier, but I was worried this would come at the cost of readability. Therefore the only place you will see IF(USE_EKF == 1) is in GM_PHD_Filter.

I've also included Ba-Ngu Vo's implementation of the Optimal Subpattern Assignment (OSPA) metric proposed by Schuhmacher et al in 
Schuhmacher, D.; Ba-Tuong Vo; Ba-Ngu Vo, "A Consistent Metric for Performance Evaluation of Multi-Object Filters," Signal Processing, IEEE Transactions on , vol.56, no.8, pp.3447,3457, Aug. 2008
This uses
- ospa_dist by Ba-Ngu Vo, taken from http://ba-ngu.vo-au.com/vo/OSPA_for_Tracks.zip
- Hungarian by Alex Melin, also taken from http://ba-ngu.vo-au.com/vo/OSPA_for_Tracks.zip

The files are mostly implemented as scripts. The main file is GM_PHD_Filter. The structure is pretty straightforward, as it mostly just calls the other scripts in a big loop.

For the linear Kalman filter, there are two files that initialise everything - GM_PHD_Initialise and GM_PHD_Simulate_Initialise. For the extended Kalman filter, there is GM_PHD_Initialisation, GM_EKF_PHD_Simulate_Initialise, and GM_EKF_PHD_Initialise_Jacobains. Changing values in these will change how the filter runs.

The other files are pretty easy to follow if you have a look in GM_PHD_Filter.m

There isn't much output on the console by default, most of it is in the graphs. Have a read of GM_PHD_Simulate_Plot to understand and edit what the different coloured dots mean. In summary, black X's are measurements, black X's with a circle around them are measurements corresponding to true targets, red/green/blue dots are true target positions, magenta circles are tracked targets, cyan triangles are the targets at the current filter iteration. If you set DRAW_VELOCITIES to 1 in GM_PHD_Initialisation, velocity arrows will be drawn on. The arrow shows direction, the length indicates magnitude (long arrow = high speed).

To see more filter output, set VERBOSE to 1 in GM_PHD_Initialisation. I can't remember exactly what this prints out, but it's more information about the inner working of the filter, and you can use it to add your own output statements wherever you want. I cut most of mine out because they were cluttering up the code.

To see the performance metric, set CALCULATE_OSPA_METRIC to 1 in GM_PHD_Initialisation. It appears as an extra line graph. High values on the line graph = high error value.

See license.txt for licensing information.

See ReleaseNotes.txt for changes between versions.

===A few remarks on the Kalman filter vs the extended Kalman filter===
If you've gotten this far I'm going to assume you understand what a KF and EKF are, so I am going to focus on the specifics of my implementation. The extended Kalman filter uses Jacobians to linearise the nonlinear prediction/observation models. These Jacobians can be calculated analytically (by hand or using a computer program; I use Wolfram Alpha as it's very easy to use) or numerically (very easy to code up but it runs much slower). I've included code for both but commented out the ones not being used. You can play around with or replace them as you see fit; see GM_EKF_PHD_Initialise_Jacobians and Test_Jacobian_Calculation.

The nonlinear observation model h() is used to map from the state space [x; y; vx; vy] to the measurement space [r; theta], primarily to generate expected observations of a target at a certain position. We also need to have an inverse observation inv_h() to perform the reverse, mapping from the measurement space [r; theta] to the state space [x; y; vx; vy] (but since we don't observe the velocity, it doesn't output velocity information and this has to be added later or not at all). The observation model is used in GM_EKF_PHD_Simulate_Measurements and GM_EKF_PHD_Construct_Update_Components. The inverse observation model is used in GM_EKF_PHD_Update, GM_EKF_PHD_Simulate_Plot and GM_PHD_Create_Birth.

If you want to replace these with different observation models, you're going to have to change these occurrences as well as the respective Jacobians.

I haven't implemented the EKF simulation given in Vo & Ma; both the EKF and KF run on the KF simulation. The prediction covariance Q given for this works well for the linear KF but not for the EKF. I set Q to eye(4) and it works; this value could probably use some further investigation and tweaking but this code is only a simulation, and a real application will probably have more grounded information on what values to set here.

Just in case someone doesn't want to use the EKF version, I'm including a zip file containing the linear-Kalman-filter-only version of the GM-PHD (version 1.05).

-- Bryan Clarke, b.clarke@acfr.usyd.edu.au

ReleaseNotes.txt
Version 1.10, 7th January 2014
==Bug Fixes==
Bug fixes in GM_PHD_Create_Birth
As pointed out by Hannes Pessentheiner, in GM_PHD_Create_Birth the weight of the existing target was not being used to scale the weight of spawned targets in the static-spawn case. This is fixed now. 
________________________________________________________________________________________________________________
Version 1.09, 13th December 2013
== Bug Fixes==
Ba-Ngu Vo has kindly allowed me to use his OSPA calculation code ospa_dist.m so CalculateOSPAMetric has been removed and replaced with ospa_dist.m. This uses Hungarian.m by Alex Melin, to whom I am also much obliged. This code is also available online at http://ba-ngu.vo-au.com/vo/OSPA_for_Tracks.zip
Apologies for putting up CalculateOSPAMetric while it was still buggy, I'll rush less in future.

I've also made this change to the linear-KF-only version 1.05 that is also included in the zip, and documented it in the respective GM_PHD_Filter.m. For versioning convenience it's now version 1.05a. I hope that this is the last update I need to make to it.
________________________________________________________________________________________________________________
Version 1.08, 12th December 2013
==Bug Fixes==
These fixes focus on the OSPA metric calculation. It's much less buggy now than it was before, but still not entirely correct. Use with caution. Hat tip to Abdulkadir Eryildirim for identifying something amiss.

Bug fixes in GM_PHD_Calculate_Performance_Metric
Corrected it so that data is passed from simTarget3History instead of simTarget2History.

Bug fixes in CalculateOSPAMetric
Added a loop to include cost for elements of Y that are not matched in X (i.e. for (m+1:n)).
________________________________________________________________________________________________________________
Version 1.07, 10th December 2013
==Bug Fixes==
Bug fixes in GM_PHD_Create_Birth and GM_EKF_PHD_Create_Birth
Restructured so that static targets will be added from the first measurement, not waiting until the second like the moving targets.
________________________________________________________________________________________________________________
Version 1.06, 2nd December 2013
==Bug Fixes==
Increased the speed of target 3 so that it more closesly matches the trajectory in Vo & Ma's example. It's still not perfect (it's a straight line rather than curving) but it's better than it was before.

==New Features==
Extended Kalman Filter
I have added an implementation of the extended Kalman filter algorithm from Vo & Ma's paper, for use with a simulated range-bearing sensor. The simulation is still the same, with a simulated range-bearing sensor pointing along the X-axis at the origin. The linear KF and EKF can be switched between by setting/clearing USE_EKF in GM_PHD_Initialisation. This will control which scripts are called in GM_PHD_Filter. The structure of the algorithm is still the same.

Just in case someone doesn't want to use the EKF and wants just the linear version, I'll include a zipped version of the linear-Kalman-filter-only GM-PHD filter.

Velocity arrows
Velocity arrows can be drawn on the simulated target plot using quiver() in both linear and EKF. To turn this on and off, set/clear DRAW_VELOCITIES in GM_PHD_Initialisation.
________________________________________________________________________________________________________________
Version 1.05, 21 November 2013
Hat-tip to Abdulkadir Eryildirim and Kevin Nickels for identifying bugs and missing features in the code.

==Bug Fixes==
Bug fixes in GM_PHD_Predict_Birth
Fixed a few comments and a print statement that had birth and spawn mixed up; this would have caused crashes when VERBOSE was set to 1.
Changed variable v in GM_PHD_Create_Birth to thisV in case I ever use a variable named v elsewhere.

Bug fixes in GM_PHD_Update
Changed variable V in GM_PHD_Update to thisV. V was already being used for the monitored volume. This may have been degrading filter performance.

Simulation plots
Removed some extra plots from GM_PHD_Simulate_Initialise and GM_PHD_Simulate_Plot; I haven't kept them updated and they're not that useful.
simTargetState3 was being predicted forward twice in GM_PHD_Simulate_Measurement. Now it is only happening once. All this means is that it moves at half the speed it did previously.

==New Features==
State Estimation in GM_PHD_Estimate
Added in the extra-display in GM_PHD_Estimate to better match Vo & Ma's algorithm. I'm not sure whether this is really necessary; it doesn't change filter performance, just output, but I want to match the pseudocode. This is controlled by setting the control variable OUTPUT_MULTIPLE_HIGH_WEIGHT_TARGETS in GM_PHD_Initialisation to 1. VERBOSE also needs to be set to 1 to see the output.

Optimal Subpattern Assignment (OSPA) metric implementation
I have added an implementation of the OSPA metric described by Schuhmacher et al in
"A Consistent Metric for Performance Evaluation of Multi-Object Filters," Signal Processing, IEEE Transactions on , vol.56, no.8, pp.3447,3457, Aug. 2008
It's not necessary for the Vo & MA PHD filter to work but provides a nice way of visualising filter performance. It's enabled in GM_PHD_Initialisation by setting CALCULATE_OSPA_METRIC to 1. The calculated metric appears as a line graph. Be advised, it slows down the filter.

==Comments==

GM_PHD_Filter.m
%GM_PHD_Filter
%Version 1.10, last modified 7th January 2014
%Matlab code by Bryan Clarke b.clarke@acfr.usyd.edu.au 
%With:
%- some Kalman filter update code by Tim Bailey, taken from his website http://www-personal.acfr.usyd.edu.au/tbailey/software/
%- error_ellipse by AJ Johnson, taken from Matlab Central http://www.mathworks.com.au/matlabcentral/fileexchange/4705-errorellipse

%The GM-PHD algorithm is from Ba-Ngu Vo & Wing-Kin Ma in:
%B.-N. Vo, W.-K. Ma, "The Gaussian Mixture Probability Hypothesis Density
%Filter", IEEE Transactions on Signal Processing, Vol 54, No. 11, November 2006, pp4091-4104

%I have implemented both the linear Kalman filter and extended Kalman
%filter versions of this algorithm; switch between these by
%setting/clearing the variable USE_EKF in GM_PHD_Initialisation.

%Ba-Ngu Vo has kindly allowed me to include his implementation of the Optimal Subpattern Assignment
%(OSPA) metric proposed by D. Schuhmacher, Ba-Tuong Vo & Ba-Ngu Vo in
% Schuhmacher, D.; Ba-Tuong Vo; Ba-Ngu Vo, "A Consistent Metric for Performance Evaluation of Multi-Object Filters," Signal Processing, IEEE Transactions on , vol.56, no.8, pp.3447,3457, Aug. 2008
%This uses
%- ospa_dist by Ba-Ngu Vo, taken from http://ba-ngu.vo-au.com/vo/OSPA_for_Tracks.zip
%- Hungarian by Alex Melin (to whom I am also much obliged), also taken from http://ba-ngu.vo-au.com/vo/OSPA_for_Tracks.zip

%The OSPA metric is not essential for the functioning of the filter but
%provides a nice way of analysing performance.

%See the README.txt, the comments and the Vo & Ma paper for more
%information about what this code actually does.

clear all;
close all;
clc;

%Step 0: Initialisation
%The EKF version must initialise the Jacobian functions used to linearise
%the observation/prediction models. The simulated measurements are also very
%different; the linear KF is direct observations of target state, the EKF is
%range-bearing measurements. The observation and prediction covariances are
%also different.
GM_PHD_Initialisation;
if USE_EKF == 0
    GM_PHD_Simulate_Initialise;
else
    GM_EKF_PHD_Simulate_Initialise;
    GM_EKF_PHD_Initialise_Jacobians;
end

%In Vo&Ma, the targets are known at filter initialisation.
%If we want to know about them, set KNOWN_TARGET to 1 in GM_PHD_Initialisation.
%Otherwise they should be initialised after being detected a few times.
%HOWEVER this is not guaranteed - sometimes due to noise or missed
%detections, one or both of the targets will not be tracked. This is just 
%part of the filter and the birth_intensity function.
if KNOWN_TARGET == 1
    t1start = [simTarget1Start(1:2); simTarget1Vel];
    t2start = [simTarget2Start(1:2); simTarget2Vel];
    m_birth = [t1start, t2start];
    w_birth = [birth_intensity(t1start), birth_intensity(t2start)];
    P_birth = [covariance_birth, covariance_birth];
    numBirthedTargets = 2;
end

%Main loop
while (k < endTime)%k = timestep
    k = k + 1;
    s = sprintf('======ITERATION %d======', k);
    disp(s);
        
    %Step Sim: Generate sensor Measurements
    %If you want to use this code with your own data or for a different problem,
    %replace this function with your own.
    if USE_EKF == 0
        GM_PHD_Simulate_Measurements;  %Linear KF measurements are simulated direct observations [X; Y] of the target positions
    else
        GM_EKF_PHD_Simulate_Measurements; %EKF measurements are simuated range-bearing [r; theta] of the target position from a fixed sensor.
    end
    
    %Linear KF use fixed matrices for prediction and update.
    %EKF calculates Jacobians to linearise the prediction and observation
    %models. There is need for an inverse sensor model to map from sensor
    %to target space for some calculations.
    if(USE_EKF == 0)
        %Step 1: Prediction for birthed/spawned targets 
        GM_PHD_Predict_Birth; 
        %Step 2: Prediction for existing targets
        GM_PHD_Predict_Existing;
        %Step 3: Construction of PHD update components
        GM_PHD_Construct_Update_Components;
        %Step 4: Update targets with measurements
        GM_PHD_Update;
    else
        %Step 1: Prediction for birthed/spawned targets 
        GM_EKF_PHD_Predict_Birth; %EKF prediction uses Jacobians to linearise the prediction model and predict the targets forward in time
        %Step 2: Prediction for existing targets
        GM_EKF_PHD_Predict_Existing;
        %Step 3: Construction of PHD update components
        GM_EKF_PHD_Construct_Update_Components;
        %Step 4: Update targets with measurements
        GM_EKF_PHD_Update;
    end
    
    %Step 5: Prune targets
    GM_PHD_Prune;
    %Step 6: Estimate position of targets
    GM_PHD_Estimate

    %Step 7: Create birthed-targets-list to add next iteration in Step 1.
    %Not a formal part of Vo&Ma but an essential step!
    %The EKF version uses an inverse sensor model.
    if(USE_EKF == 0)
        GM_PHD_Create_Birth; 
    else
        GM_EKF_PHD_Create_Birth;
    end
    
    %Step Metric: Calculate performance metric
    GM_PHD_Calculate_Performance_Metric;
    
    %Step Plot: Generate graphs
    %The EKF version uses an inverse sensor model.
    if USE_EKF == 0
        GM_PHD_Simulate_Plot;
    else
       GM_EKF_PHD_Simulate_Plot;
    end
    
    if(VERBOSE == true)
        pause;%Pause to allow reading of the text
    end

end

